File: server/utils/extraHelpers/postgresqlHelper.js
'use strict';

/* 3rd-party Modules */
var pg     = require('pg');
var moment = require('moment-timezone');

/* Project Modules */
var CONFIG            = require('../yamlResources').get('CONFIG');
var toolkit           = require('../toolkit');
var logHelper         = require('../logHelper');
var PostgreSQLBuilder = require('./sqlBuilder').PostgreSQLBuilder;

// Use parseInt to parse INT8
var pgTypes = pg.types;
pgTypes.setTypeParser(pgTypes.builtins.INT8, function(v) {
  return parseInt(v, 10);
});

function getConfig(c) {
  return {
    host                   : c.host,
    port                   : c.port,
    user                   : c.user,
    password               : c.password,
    database               : c.database,
    max                    : CONFIG._DB_POOL_SIZE_SERVER,
    connectionTimeoutMillis: CONFIG._DB_CONN_TIMEOUT * 1000,
  };
};

/* Singleton Client */
var CLIENT_CONFIG = null;
var CLIENT        = null;

/**
 * @constructor
 * @param  {Object}  [logger=null]
 * @param  {Object}  [config=null]
 * @param  {Boolean} [debug=false]
 * @return {Object} - MySQL Helper
 */
var PostgreSQLHelper = function(logger, config, debug) {
  this.logger = logger || logHelper.createHelper();

  this.isDryRun = false;
  this.skipLog  = false;

  // TODO PK conflict notice
  this.uniqueKeyErrorKeyword = 'duplicate key value';
  this.uniqueKeyErrorRegExp  = /Key \'(.+)\'/;

  if (config) {
    this.config = toolkit.noNullOrWhiteSpace(config);
    this.client = new pg.Pool(getConfig(this.config));

  } else {
    if (!CLIENT) {
      CLIENT_CONFIG = toolkit.noNullOrWhiteSpace({
        host    : CONFIG.POSTGRESQL_HOST,
        port    : CONFIG.POSTGRESQL_PORT,
        user    : CONFIG.POSTGRESQL_USER,
        password: CONFIG.POSTGRESQL_PASSWORD,
        database: CONFIG.POSTGRESQL_DATABASE,
      });
      CLIENT = new pg.Pool(getConfig(CLIENT_CONFIG));
    }

    this.config = CLIENT_CONFIG;
    this.client = CLIENT;
  }

  this.transConn = null;
  this.transConnRing = 0;
};

/**
 * Close the DB connection
 */
PostgreSQLHelper.prototype.close = function() {
  // Do not close the default DB connection/pool
  if (CLIENT === this.client) return;

  this.client.end();
};

PostgreSQLHelper.prototype.getTimezone = function() {
  return this.config.timezone || CONFIG['TIMEZONE'];
};

PostgreSQLHelper.prototype.version = function(callback) {
  var self = this;

  var version = null;
  var branch  = null;

  // Get version
  var sql = self.createSQLBuilder(`SHOW server_version`);
  self.query(sql, function(err, dbRes) {
    if (err) return callback(err);

    var versionStr = dbRes[0].server_version;
    var m = versionStr.match(/\d+\.\d+/)
    if (m) {
      version = m[0];
    } else {
      version = versionStr;
    }

    // TODO Get branch
    branch = 'PostgreSQL';

    return callback(null, version, branch);
  });
};

PostgreSQLHelper.prototype.settings = function(callback) {
  var sql = this.createSQLBuilder(`SHOW ALL`);
  this.query(sql, function(err, dbRes) {
    if (err) return callback(err);

    var settings = {};
    dbRes.forEach(function(d) {
      var k = d['name'];
      var v = d['setting'];
      settings[k] = v;
    });

    return callback(null, settings);
  });
};

PostgreSQLHelper.prototype.tables = function(callback) {
  var sql = this.createSQLBuilder(`
    SELECT
      relname AS name
    FROM
      pg_stat_user_tables
    ORDER BY
      relname ASC`);
  this.query(sql, function(err, dbRes) {
    if (err) return callback(err);

    var tables = dbRes.reduce(function(acc, x) {
      acc.push(x.name);
      return acc;
    }, []);

    return callback(null, tables);
  });
};

PostgreSQLHelper.prototype.columns = function(table, callback) {
  var sql = this.createSQLBuilder(`
    SELECT
      column_name AS name
    FROM
      information_schema.columns
    WHERE
      table_schema   = 'public'
      AND table_name = ?`);
  this.query(sql, [ table ], function(err, dbRes) {
    if (err) return callback(err);

    var columns = dbRes.reduce(function(acc, x) {
      acc.push(x.name);
      return acc;
    }, []);

    return callback(null, columns);
  });
};

PostgreSQLHelper.prototype.tableStatus = function(callback) {
  var sql = this.createSQLBuilder(`
    SELECT
      relname    AS name,
      n_live_tup AS rows,

      pg_table_size(relid)   AS data_length,
      pg_indexes_size(relid) AS index_length
    FROM
      pg_stat_user_tables
    ORDER BY
      relname ASC
  `);
  this.query(sql, function(err, dbRes) {
    if (err) return callback(err);

    var tableStatus = dbRes.reduce(function(acc, x) {
      var totalSize  = (x.data_length || 0) + (x.index_length || 0);
      var avgRowSize = x.rows ? parseInt(totalSize / x.rows) : 0;

      var t = {
        name: x.name,
        rows: x.rows || 0,

        dataSize  : x.data_length  || 0,
        indexSize : x.index_length || 0,
        totalSize : totalSize,
        avgRowSize: avgRowSize,
      }

      acc[x.name] = t;
      return acc;
    }, {});

    return callback(null, tableStatus);
  });
};

/**
 * Run SQL statement.
 *
 * @param  {String}   sql       - SQL statement
 * @param  {*[]}      sqlParams - SQL parameters
 * @param  {Function} callback
 * @return {undefined}
 */
PostgreSQLHelper.prototype.query = function(sql, sqlParams, callback) {
  var self = this;

  if ('function' === typeof sqlParams) {
    callback  = sqlParams;
    sqlParams = undefined;
  }

  callback = toolkit.ensureFn(callback);

  sql = this._prepareSQL(sql, sqlParams);
  var debugSQL = toolkit.toDebugText(sql);

  // Query callback
  function queryCallback(err, dbRes) {
    if (err) {
      if (!self.transConn) {
        self.logger.error('[POSTGRESQL] Query: {0}', debugSQL);
      } else {
        self.logger.error('[POSTGRESQL] Trans Query: {0}', debugSQL);
      }
      self.logger.error(`[POSTGRESQL ERROR] ${err}`);

      return callback(err);
    }

    if (toolkit.startsWith(sql.toUpperCase(), 'SELECT')) {
      // Do type conversion for SELECT query
      if (Array.isArray(dbRes)) {
        // Only convert value type in first dataset when got multiple dataset
        dbRes = dbRes.map(function(d) {
          return d.rows;
        });

        self._convertTypes(dbRes[0]);
      } else {
        // Convert value type
        dbRes = dbRes.rows;

        self._convertTypes(dbRes);
      }
    } else {
      // Retrun row counts for other query
      dbRes = dbRes.rows;
    }

    return callback(null, dbRes);
  }

  if (!self.transConn) {
    if (!self.skipLog) {
      self.logger.debug('[POSTGRESQL] Query: {0}', debugSQL);
    }

    // Single Query
    if (!self.isDryRun || sql.trim().indexOf('SELECT') === 0) {
      return self.client.query(sql, queryCallback);
    } else {
      return queryCallback();
    }

  } else {
    // Transaction Query
    if (!self.skipLog) {
      self.logger.debug('[POSTGRESQL] Trans Query: {0}', debugSQL);
    }

    return self.transConn.query(sql, queryCallback);
  }
};

/**
 * Get a connection from connection pool and start a transaction.
 *
 * @param  {Function} callback
 * @return {undefined}
 */
PostgreSQLHelper.prototype.startTrans = function(callback) {
  var self = this;
  callback = toolkit.ensureFn(callback);

  if (self.transConn) {
    self.transConnRing++;

    if (!self.skipLog) {
      self.logger.debug('[POSTGRESQL] Enter ring {0} -> {1} (by START)',
        self.transConnRing - 1,
        self.transConnRing
      );
    }

    return callback();
  }

  self.client.connect(function(err, conn) {
    if (err) return callback(err);

    conn.query('BEGIN', function(err) {
      if (err) {
        conn.release()

        return callback(err);

      } else {
        self.transConn = conn;
        self.transConnRing++;

        if (!self.skipLog) {
          self.logger.debug('[POSTGRESQL] Trans START');
        }

        return callback();
      }
    });
  });
};

/**
 * Commit the transaction.
 *
 * @param  {Function} callback
 * @return {undefined}
 */
PostgreSQLHelper.prototype.commit = function(callback) {
  var self = this;
  callback = toolkit.ensureFn(callback);

  if (self.isDryRun) return self.rollback(callback);

  if (!self.transConn) {
    if (!self.skipLog) {
      self.logger.warning('[POSTGRESQL] Trans COMMIT (Transaction not started, skip)');
    }
    return callback();

  } else {
    self.transConnRing--;

    if (self.transConnRing <= 0) {
      self.transConn.query('COMMIT', function(err) {
        self.transConn.release();
        self.transConn = null;

        if (!self.skipLog) {
          self.logger.debug('[POSTGRESQL] Trans COMMIT');
        }

        return callback(err);
      });

    } else {
      if (!self.skipLog) {
        self.logger.debug('[POSTGRESQL] Leave ring {0} <- {1} (by COMMIT)',
          self.transConnRing,
          self.transConnRing + 1
        );
      }

      return callback();
    }
  }
};

/**
 * Rollback the transaction.
 *
 * @param  {Function} callback
 * @return {undefined}
 */
PostgreSQLHelper.prototype.rollback = function(callback) {
  var self = this;
  callback = toolkit.ensureFn(callback);

  if (!self.transConn) {
    if (!self.skipLog) {
      self.logger.warning('[POSTGRESQL] Trans ROLLBACK (Transaction not started, skip)');
    }
    return callback();

  } else {
    self.transConnRing--;

    if (self.transConnRing <= 0) {
      self.transConn.query('ROLLBACK', function() {
        self.transConn.release();
        self.transConn = null;

        if (!self.skipLog) {
          self.logger.debug('[POSTGRESQL] Trans ROLLBACK');
        }

        return callback();
      });

    } else {
      if (!self.skipLog) {
        self.logger.debug('[POSTGRESQL] Leave ring {0} <- {1} (by ROLLBACK)',
          self.transConnRing,
          self.transConnRing + 1
        );
      }

      return callback();
    }
  }
};

PostgreSQLHelper.prototype._convertTypes = function(dbRes) {
  var timezone = this.getTimezone();

  dbRes.forEach(function(d) {
    for (var k in d) {
      var v = d[k];

      // JSON field to obj
      if (toolkit.endsWith(k, 'JSON') && 'string' === typeof v) {
        try {
          d[k] = JSON.parse(v);
        } catch(e) {
          // Nope
        }
      }

      // DateTime field to Moment obj
      if (timezone && toolkit.endsWith(k, 'Time') && 'number' === typeof v) {
        try {
          d[k] = moment(v * 1000).tz(timezone);
        } catch(e) {
          // Nope
        }
      }
    }
  });
};

PostgreSQLHelper.prototype._prepareSQL = function(sql, sqlParams) {
  var self = this;

  var sqls = toolkit.asArray(sql);
  var preparedSQL = sqls.map(function(_sql) {
    if (self.client === CLIENT) {
      // Special for system DB

      // Force to use SQLBuilder
      if (!(_sql instanceof PostgreSQLBuilder)) {
        var e = new Error(`SQL is not a instance of PostgreSQLBuilder: ${_sql}`);
        throw e;
      }

      // Check SQL safty
      _sql.checkSafty();

      // Add updateTime / createTime automatically
      var nowTimestamp = toolkit.getTimestamp();
      if (_sql.meta.type === 'INSERT') {
        _sql.UPDATE_VALUES({
          'createTime': nowTimestamp,
          'updateTime': nowTimestamp,
        }, true);

      } else if (_sql.meta.type === 'UPDATE') {
        _sql.SET({
          'updateTime': nowTimestamp,
        });
      }
    }

    // Convert to string with semicolons
    _sql = _sql.toString().trim();
    if (!toolkit.endsWith(_sql, ';')) {
      _sql += ';';
    }

    return _sql;
  }).join(' ');

  if (sqlParams) {
    preparedSQL = sqls[0].formatSQL(preparedSQL, sqlParams);
  }

  return preparedSQL;
};

/**
 * Get page infomation from DB result by paging setting.
 *
 * @param  {Object}   options - Paging options
 * @param  {Object[]} dbRes
 * @return {Object}
 */
PostgreSQLHelper.prototype.getPageInfo = function(options, dbRes) {
  if (!options) return null;

  var pageInfo = {
    pagingStyle: options.pagingStyle,
    pageSize   : options.pageSize,
  }

  if (options.pagingStyle === true || options.pagingStyle === 'normal') {
    pageInfo.count       = dbRes[0].length;
    pageInfo.totalCount  = dbRes[1][0].totalCount;
    pageInfo.pageNumber  = options.pageNumber || 1;
    pageInfo.pageCount   = Math.ceil(pageInfo.totalCount / pageInfo.pageSize);
    pageInfo.isFirstPage = pageInfo.pageNumber <= 1;

  } else if (options.pagingStyle === 'simple') {
    pageInfo.count       = dbRes.length;
    pageInfo.pageNumber  = options.pageNumber || 1;
    pageInfo.isFirstPage = pageInfo.pageNumber <= 1;

  } else if (options.pagingStyle === 'marker') {
    pageInfo.count           = dbRes.length;
    pageInfo.pageMarkerField = null;
    pageInfo.pageMarker      = null;
    pageInfo.isFirstPage     = null;

    if (dbRes.length > 0 && options.pageMarkerField) {
      pageInfo.pageMarkerField = options.pageMarkerField.split('.').pop();
      pageInfo.pageMarker      = dbRes[dbRes.length - 1][pageInfo.pageMarkerField];
    }

    pageInfo.isFirstPage = options.pageMarker <= 0;
  }

  return pageInfo;
};

/**
 * Get data from DB result when paging is setted.
 *
 * @param  {Object}   options - Paging options
 * @param  {Object[]} dbRes
 * @return {Object}
 */
PostgreSQLHelper.prototype.getData = function(options, dbRes) {
  if (!options) return dbRes;

  if (options.pagingStyle === true || options.pagingStyle === 'normal') {
    return dbRes[0];
  } else {
    return dbRes;
  }
};

PostgreSQLHelper.prototype.createSQLBuilder = function(rawSQL) {
  return new PostgreSQLBuilder(rawSQL).setTimezone(this.getTimezone());
};

exports.PostgreSQLHelper = PostgreSQLHelper;

exports.createHelper = function(logger, config, debug) {
  return new PostgreSQLHelper(logger, config, debug);
};

File: worker/utils/extra_helpers/postgresql_helper.py
# -*- coding: utf-8 -*-

# Built-in Modules
import re
import time
import traceback

# 3rd-party Modules
import arrow
import psycopg2
from dbutils.pooled_db import PooledDB
import sqlparse

# Project Modules
from worker.utils import toolkit, yaml_resources
from worker.utils.extra_helpers import format_sql, postgresql_escape, table_to_guance_dql_like_result
from worker.utils.extra_helpers import to_dict_rows
from worker.utils.extra_helpers.sql_builder import PostgreSQLBuilder

CONFIG = yaml_resources.get('CONFIG')

def get_config(c):
    _encoding = c.get('encoding') or c.get('charset') or 'utf8'

    config = {
        'host'           : c.get('host') or '127.0.0.1',
        'port'           : c.get('port') or 5432,
        'user'           : c.get('user'),
        'password'       : c.get('password'),
        'dbname'         : c.get('dbname') or c.get('database'),
        'client_encoding': _encoding,

        'connect_timeout': CONFIG['_DB_CONN_TIMEOUT'],
        'maxusage'       : CONFIG['_DB_POOL_MAX_USAGE'],
        'maxconnections': c.get('maxconnections') or 1,
    }
    return config

CLIENT_CREATE_TIME = 0
CLIENT_CONFIG      = None
CLIENT             = None

class FuncPostgreSQLHelper(object):
    db_name = 'PostgreSQL'

    def __init__(self, logger, config=None, database=None, pool_size=None, *args, **kwargs):
        self.logger = logger

        self.skip_log = False

        if config:
            if database:
                config['database'] = database

            if pool_size:
                config['maxconnections'] = pool_size

            self.config = config
            self.client = PooledDB(psycopg2, **get_config(config))

        else:
            global CLIENT_CREATE_TIME
            global CLIENT_CONFIG
            global CLIENT

            if CLIENT and time.time() - CLIENT_CREATE_TIME > CONFIG['_DB_POOL_RECYCLE_TIMEOUT']:
                # TODO Close the connection pool?
                CLIENT.close()

                CLIENT_CREATE_TIME = None
                CLIENT             = None

            if not CLIENT:
                CLIENT_CREATE_TIME = time.time()
                CLIENT_CONFIG = {
                    'host'          : CONFIG['POSTGRESQL_HOST'],
                    'port'          : CONFIG['POSTGRESQL_PORT'],
                    'user'          : CONFIG['POSTGRESQL_USER'],
                    'password'      : CONFIG['POSTGRESQL_PASSWORD'],
                    'database'      : CONFIG['POSTGRESQL_DATABASE'],
                    'encoding'      : CONFIG['_POSTGRESQL_ENCODING'],
                    'maxconnections': CONFIG['_DB_POOL_SIZE_WORKER'],
                }
                CLIENT = PooledDB(psycopg2, **get_config(CLIENT_CONFIG))

            self.config = CLIENT_CONFIG
            self.client = CLIENT

    def __del__(self):
        # Do not close the default DB connection/pool
        if not self.client or self.client is CLIENT or not isinstance(self.client, PooledDB):
            return

        try:
            self.client.close()

        except Exception as e:
            for line in traceback.format_exc().splitlines():
                self.logger.error(line)

        finally:
            self.client = None

    @classmethod
    def format_sql(cls, sql, sql_params=None, pretty=False):
        return format_sql(sql, sql_params, pretty, sql_escape=postgresql_escape)

    @property
    def timezone(self):
        return self.config.get('timezone') or CONFIG.get('TIMEZONE')

    def check(self):
        try:
            sql = self.create_sql_builder('SELECT 1')
            self.query(sql)

        except Exception as e:
            for line in traceback.format_exc().splitlines():
                self.logger.error(line)

            raise

    def tables(self):
        sql = self.create_sql_builder('''
            SELECT
                relname AS name
            FROM
                pg_stat_user_tables
            ORDER BY
                relname ASC
        ''')

        db_res = self.query(sql)

        tables = list(map(lambda x: x['name'], db_res))
        return tables

    def table_status(self):
        sql = self.create_sql_builder('''
            SELECT
                relname    AS name,
                n_live_tup AS rows,

                pg_table_size(relid)   AS data_length,
                pg_indexes_size(relid) AS index_length
            FROM
                pg_stat_user_tables
            ORDER BY
                relname ASC
        ''')

        db_res = self.query(sql)

        table_status = {}
        for d in db_res:
            total_size   = (d['data_length'] or 0) + (d['index_length'] or 0)
            avg_row_size = int(total_size / d['rows']) if d.get('rows') else 0

            t = {
                'name'   : d['name'],
                'rows'   : d.get('rows') or 0,

                'dataSize'  : d.get('data_length')  or 0,
                'indexSize' : d.get('index_length') or 0,
                'totalSize' : total_size,
                'avgRowSize': avg_row_size,
            }

            table_status[d['name']] = t

        return table_status

    def clear_table(self, table):
        # Check if table exists
        sql = self.create_sql_builder('''
            SELECT
                relname
            FROM
                pg_stat_user_tables
            WHERE
                relname = ?
        ''')
        sql_params = [ table ]

        db_res = self.query(sql, sql_params)
        if not db_res:
            self.logger.debug(f'[POSTGRESQL] The table `{table}` is not exist, skip clearing')
            return

        # Check if data exists
        sql = self.create_sql_builder('SELECT * FROM ?? LIMIT 1')
        sql_params = [ table ]

        has_data = bool(self.query(sql, sql_params))
        if not has_data:
            self.logger.debug(f'[POSTGRESQL] The table `{table}` has no data, skip clearing')
            return

        # Do truncate
        sql = self.create_sql_builder()
        sql.TRUNCATE(table)

        self.query(sql)

    def start_trans(self):
        if not self.skip_log:
            self.logger.debug('[POSTGRESQL] Trans START')

        conn = self.client.connection()
        cur  = conn.cursor()

        trans_conn = {
            'conn': conn,
            'cur' : cur,
        }

        return trans_conn

    def commit(self, trans_conn):
        if not trans_conn:
            return

        if not self.skip_log:
            self.logger.debug('[POSTGRESQL] Trans COMMIT')

        conn = trans_conn.get('conn')
        cur  = trans_conn.get('cur')

        conn.commit()

        cur.close()
        conn.close()

    def rollback(self, trans_conn):
        if not trans_conn:
            return

        if not self.skip_log:
            self.logger.debug('[POSTGRESQL] Trans ROLLBACK')

        conn = trans_conn.get('conn')
        cur  = trans_conn.get('cur')

        conn.rollback()

        cur.close()
        conn.close()

    def _convert_types(self, db_res):
        for d in db_res:
            for k, v in d.items():
                # JSON Fields to Objects
                if k.endswith('JSON') and isinstance(v, str):
                    try:
                        d[k] = toolkit.json_loads(v)
                    except Exception as e:
                        pass

                # Time Fields to Arrow Objects
                if self.timezone and k.endswith('Time') and isinstance(v, (int, float)):
                    try:
                        d[k] = arrow.get(v, self.timezone)
                    except Exception as e:
                        pass

        return db_res

    def _prepare_sql(self, sql):
        sqls = []
        for _sql in toolkit.as_array(sql):
            if self.client is CLIENT:
                # Special processing for system DB

                # Force using SQLBuilder
                if not isinstance(_sql, PostgreSQLBuilder):
                    e = Exception(f'SQL is not a instance of PostgreSQLBuilder: {_sql}')
                    raise e

                # Check SQL security
                _sql.check_safty()

                # Add updateTime / createTime automatically
                now_timestamp = toolkit.get_timestamp()
                if _sql.meta['type'] == 'INSERT':
                    _sql.UPDATE_VALUES({
                        'createTime': now_timestamp,
                        'updateTime': now_timestamp,
                    }, True)

                elif _sql.meta['type'] == 'UPDATE':
                    _sql.SET({
                        'updateTime': now_timestamp,
                    })

            # Convert to string and add semicolon
            _sql = str(_sql).strip()
            if not _sql.endswith(';'):
                _sql += ';'

            sqls.append(_sql)

        return ' '.join(sqls)

    def _trans_execute(self, trans_conn, sql, sql_params=None):
        sql       = self._prepare_sql(sql)
        sql       = self.format_sql(sql, sql_params)
        debug_sql = toolkit.to_debug_text(sql)

        if not trans_conn:
            raise Exception('Transaction not started')

        conn = trans_conn['conn']
        cur  = trans_conn['cur']

        try:
            dt = toolkit.DiffTimer()

            cur.execute(sql)
            count = cur.rowcount

            db_res = None
            try:
                db_res = cur.fetchall()
            except psycopg2.ProgrammingError as e:
                if str(e) == 'no results to fetch':
                    pass
                else:
                    raise

            if not self.skip_log:
                self.logger.debug(f'[POSTGRESQL] Trans Query `{debug_sql}` (Cost: {dt.tick()} ms)')

            db_res = list(db_res or [])
            if db_res:
                db_res = to_dict_rows(cur, db_res)
                db_res = self._convert_types(db_res)

            return db_res, count

        except Exception as e:
            self.logger.error(f'[POSTGRESQL] Trans Query `{debug_sql}` (Cost: {dt.tick()} ms)')
            raise

    def _execute(self, sql, sql_params=None):
        sql       = self._prepare_sql(sql)
        sql       = self.format_sql(sql, sql_params)
        debug_sql = toolkit.to_debug_text(sql)

        conn = None
        cur  = None

        try:
            dt = toolkit.DiffTimer()

            conn = self.client.connection()
            cur  = conn.cursor()

            cur.execute(sql)
            count = cur.rowcount

            db_res = None
            try:
                db_res = cur.fetchall()
            except psycopg2.ProgrammingError as e:
                if str(e) == 'no results to fetch':
                    pass
                else:
                    raise

            conn.commit()

            if not self.skip_log:
                self.logger.debug(f'[POSTGRESQL] Query `{debug_sql}` (Cost: {dt.tick()} ms)')

            db_res = list(db_res or [])
            if db_res:
                db_res = to_dict_rows(cur, db_res)
                db_res = self._convert_types(db_res)

            return db_res, count

        except Exception as e:
            for line in traceback.format_exc().splitlines():
                self.logger.error(line)

            if conn:
                conn.rollback()

            self.logger.error(f'[POSTGRESQL] Query `{debug_sql}` (Cost: {dt.tick()} ms)')
            raise

        finally:
            if cur:
                cur.close()

            if conn:
                conn.close()

    def trans_query(self, trans_conn, sql, sql_params=None):
        result, count = self._trans_execute(trans_conn, sql, sql_params)
        return result

    def trans_non_query(self, trans_conn, sql, sql_params=None):
        result, count = self._trans_execute(trans_conn, sql, sql_params)
        return count

    def query(self, sql, sql_params=None):
        result, count = self._execute(sql, sql_params)
        return result

    def non_query(self, sql, sql_params=None):
        result, count = self._execute(sql, sql_params)
        return count

    def create_sql_builder(self, raw_sql=None):
        return PostgreSQLBuilder(raw_sql).set_timezone(self.timezone)

    def trans_query_raw(self, trans_conn, sql, sql_params=None):
        sql = self.create_sql_builder(sql)
        return self.trans_query(trans_conn, sql, sql_params)

    def trans_non_query_raw(self, trans_conn, sql, sql_params=None):
        sql = self.create_sql_builder(sql)
        return self.trans_non_query(trans_conn, sql, sql_params)
        return count

    def query_raw(self, sql, sql_params=None):
        sql = self.create_sql_builder(sql)
        return self.query(sql, sql_params)

    def non_query_raw(self, sql, sql_params=None):
        sql = self.create_sql_builder(sql)
        return self.non_query(sql, sql_params)

    def guance_dql_like_query(self, query_statement, options=None):
        options = options or {}

        # Run only the first statement.
        sql = sqlparse.split(query_statement)[0].strip().strip(';')

        # Checks if SELECT
        parsed_sql = sqlparse.parse(sql)[0]
        if parsed_sql.get_type() != 'SELECT':
            e = Exception('Only SELECT SQL is allowed in DQL-like query')
            raise e

        # SQL wrap
        sql = f'SELECT * FROM ({sql}) AS "main"'

        # Check If time field exists
        has_time_col = False
        for col in list(parsed_sql.get_sublists())[0].get_sublists():
            if col.get_name() == 'time':
                has_time_col = True
                break

        # Add time range filter (WHERE time...) only if time field exists
        if has_time_col:
            where_conditions = [
                f'"time" >= {options["start"]}',
                f'"time" < {options["end"]}',
            ]
            sql += f" WHERE {' AND '.join(where_conditions)}"

        # Add a query limit (LIMIT)
        sql += f" LIMIT {CONFIG['GUANCE_DQL_LIKE_QUERY_LIMIT']}"

        # Run SQL
        db_res = self.query(sql)

        # Convert format
        guance_dql_like_res = table_to_guance_dql_like_result(db_res)
        guance_dql_like_res['executedQueryStatement'] = sql
        return guance_dql_like_res

class PostgreSQLHelper(FuncPostgreSQLHelper):
    def _prepare_sql(self, sql):
        if isinstance(sql, str):
            sql = self.create_sql_builder(sql)

        return super()._prepare_sql(sql)
